### PR TITLE
42 dfri recursive component

### PR DESCRIPTION
**Pull request by ETatuzova**

### CODE REVIEW

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp [hunk 1]
```diff
@@ -0,0 +1,932 @@
...
+
+                std::vector<var> negative_xs;
+                for(std::size_t i=0; i< fri_params.lambda; i++){
+                    assignment.witness(component.W(0), row) = var_value(assignment, xs[i]);
+                    assignment.witness(component.W(1), row) = -var_value(assignment, xs[i]);
+                    negative_xs.push_back(var(component.W(1), row, false));
+                    row++;
+                }
+
+                std::vector<std::pair<std::size_t, std::size_t>> ordered_eval_map;
```
Comment (ETatuzova) [inline crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp:461]: Move it to constructor

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp [hunk 2]
```diff
@@ -0,0 +1,932 @@
...
+                std::sort(ordered_eval_map.begin(), ordered_eval_map.end(), [](const std::pair<std::size_t,std::size_t> &left, const std::pair<std::size_t,std::size_t> &right){
+                    return left.second < right.second || (left.second == right.second && left.first < right.first);
+                });
+
+                std::vector<std::uint32_t> witnesses;
+                for (std::uint32_t i = 0; i < component.witness_amount(); i++) {
+                    witnesses.push_back(component.W(i));
+                }
+
+                dfri_linear_check_component_type dfri_linear_check_instance = dfri_linear_check_component_type(witnesses, std::array<std::uint32_t, 1>(), std::array<std::uint32_t, 0>(), ordered_eval_map.size(), ordered_eval_map);
```
Comment (ETatuzova) [inline crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp:483]: Move component instance constructor to dFRI verifier constructor

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp [hunk 3]
```diff
@@ -0,0 +1,932 @@
...
+                        x_index_input.b.push_back(instance_input.merkle_tree_positions[i][j]);
+                    }
+                    typename x_index_component_type::result_type x_index_output = generate_assignments(
+                        x_index_instance, assignment, x_index_input, row
+                    );
+                    row += x_index_instance.rows_amount;
+                    x_indices.push_back(x_index_output.b0);
+                }
+
+                std::vector<var> negative_xs;
```
Comment (ETatuzova) [inline crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp:453]: There is no need to use lambda rows for it.
Place pairs {x, -x} horizontally

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp [hunk 4]
```diff
@@ -0,0 +1,932 @@
...
+
+                std::size_t rows = 0;
+                std::size_t poseidon_rows = 0;
+
+                const auto &poseidon_instance = component.poseidon_instance;
+                const auto &constant_pow_instance = component.constant_pow_instance;
+                const auto &constant_pow2_instance = component.constant_pow2_instance;
+                const auto &x_index_instance = component.x_index_instance;
+                const auto &colinear_checks_instance = component.colinear_checks_instance;
+                // const auto &dfri_linear_check_instance = component.dfri_linear_check_instance;
```
Comment (ETatuzova) [inline crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp:685]: Why is it commented?

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp [hunk 5]
```diff
@@ -0,0 +1,932 @@
...
+                        x_index_input.b.push_back(instance_input.merkle_tree_positions[i][j]);
+                    }
+                    typename x_index_component_type::result_type x_index_output  = generate_circuit(
+                        x_index_instance, bp, assignment, x_index_input, row
+                    );
+                    row += x_index_instance.rows_amount;
+                    x_indices.push_back(x_index_output.b0);
+                }
+
+                std::vector<var> negative_xs;
```
Comment (ETatuzova) [inline crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp:760]: Same with similar comment in generate_assignments.
Place pairs horizontally

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp [hunk 6]
```diff
@@ -0,0 +1,932 @@
...
+                for(std::size_t i=0; i< fri_params.lambda; i++){
+                    auto xs_copy = var(component.W(0), row, false);
+                    auto xs_neg  = var(component.W(1), row, false);
+                    negative_xs.push_back(xs_neg);
+                    bp.add_copy_constraint({xs_copy, xs[i]});
+                    assignment.enable_selector(selector, row);
+                    row++;
+                }
+
+                std::vector<std::pair<std::size_t, std::size_t>> ordered_eval_map;
```
Comment (ETatuzova) [inline crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp:771]: No need in code duplication. Move to constructor.

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp [hunk 7]
```diff
@@ -0,0 +1,932 @@
...
+                std::sort(ordered_eval_map.begin(), ordered_eval_map.end(), [](const std::pair<std::size_t,std::size_t> &left, const std::pair<std::size_t,std::size_t> &right){
+                    return left.second < right.second || (left.second == right.second && left.first < right.first);
+                });
+
+                std::vector<std::uint32_t> witnesses;
+                for (std::uint32_t i = 0; i < component.witness_amount(); i++) {
+                    witnesses.push_back(component.W(i));
+                }
+
+                dfri_linear_check_component_type dfri_linear_check_instance = dfri_linear_check_component_type(witnesses, std::array<std::uint32_t, 1>(), std::array<std::uint32_t, 0>(), ordered_eval_map.size(), ordered_eval_map);
```
Comment (ETatuzova) [inline crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp:793]: Move to dFRI component constructor

Comment (Valeh2012) [reply crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp]: Moved to dFRI component constructor

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp (hunk 6 above)
Comment (Valeh2012) [reply crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp]: Moved to dFRI component constructor

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp (hunk 5 above)
Comment (Valeh2012) [reply crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp]: flexible_negate component added for this operation

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp (hunk 4 above)
Comment (Valeh2012) [reply crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp]: uncommented

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp (hunk 3 above)
Comment (Valeh2012) [reply crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp]: flexible_negate component created for this part

File: crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp (hunk 1 above)
Comment (Valeh2012) [reply crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp]: Moved to dFRI component constructor

**Pull request closed without merging**